### PR TITLE
Maven

### PR DESCRIPTION
**Pull request by dwightguth**
This is going to fail on Jenkins, so we'll deal with Jenkins separately, right now I just want feedback on the diff. I suggest you review the two commits separately as they are logically isolated. The only exception is pom.xml.

### CODE REVIEW

**Comment by dwightguth** [comment]
@andreistefanescu @cos @radumereuta @yilongli @bmmoore Please review. Anyone else reading this should review as well because this diff affects **YOUR** build system.

**Comment by dwightguth** [comment]
@pdaian Please review with your experience with maven also.

**Comment by cos** [comment]
Maven doesn't have the most concise syntax. Could you please explain in plain english what you did?

**Comment by dwightguth** [comment]
- made the build platform dependent so that users will download a version of the framework specific to their os and architecture
- moved a lot of files around, chiefly:
  - sdf code is now in src/main/sdf, and is copied to target/sdf and built there using an ant plugin
  - binary files now live as artifacts in the maven repo and are downloaded based on your architecture and os
  - properties files now live as resources in the jar instead of in lib and include
  - jj files now live in src/main/javacc
  - java files now live in src/main/java
  - unit and integration tests now live in src/test/java
  - files used by tests now live in src/test/resources
  - the version information now lives in the jar manifest
  - all binaries except maude now live in the bin directory inside the release
  - all native libraries loaded via System.loadLibrary now live in lib/native inside the release
  - scripts are now platform dependent and live in the bin and lib directories in the release
- maven profiles are used to select an architecture and os to build for.
- ktest now copies its outputted junit reports to a directory inside the target, which should hypothetically be readable by jenkins
- added some more unit test coverage in places
- minor changes to the code to support the new location for various files. 

Is that enough for you? If you look at the pom.xml and have questions about more specific sections I can answer those as well.

**Comment by dwightguth** [comment]
For what it's worth I am open to suggestions also for things to change about how this is structured, I just moved all the files to the places I thought were most appropriate for them, given what they do. But one of the things I'm hoping for is feedback on how to structure this better.

**Comment by radumereuta** [comment]
I've checkout this branch, but crashed at mvn package with this error:

```
Results :

Tests in error:
  UseSMTTest.testGetModel:44 » UnsatisfiedLink D:\work\k\target\dependency\z3jav...

Tests run: 61, Failures: 0, Errors: 1, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 01:40 min
[INFO] Finished at: 2014-08-13T13:36:15-05:00
[INFO] Final Memory: 32M/272M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.17:test (default-test) on project k: There are test failures.
[ERROR]
[ERROR] Please refer to d:\work\k\target\surefire-reports for the individual test results.
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
d:\work\k>ant
Buildfile: d:\work\k\build.xml

BUILD FAILED
Target "build" does not exist in the project "KTool Build file".
```

I have no idea if this is helpful or not. So ask for more information if needed.
Platform used for test: Win7x64

**Comment by dwightguth** [comment]
Yeah I need the actual stack trace from the failed test, it's in the target/surefire-reports directory

**Comment by radumereuta** [comment]
```
-------------------------------------------------------------------------------
Test set: org.kframework.backend.java.symbolic.UseSMTTest
-------------------------------------------------------------------------------
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.09 sec <<< FAILURE! - in org.kframework.backend.java.symbolic.UseSMTTest
testGetModel(org.kframework.backend.java.symbolic.UseSMTTest)  Time elapsed: 0.04 sec  <<< ERROR!
java.lang.UnsatisfiedLinkError: D:\work\k - Copy\target\dependency\z3java.dll: Can't find dependent libraries
    at java.lang.ClassLoader$NativeLibrary.load(Native Method)
    at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1929)
    at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1847)
    at java.lang.Runtime.loadLibrary0(Runtime.java:870)
    at java.lang.System.loadLibrary(System.java:1119)
    at com.microsoft.z3.Native.<clinit>(Native.java:10)
    at com.microsoft.z3.Context.<init>(Context.java:23)
    at org.kframework.backend.java.symbolic.UseSMT.checkSat(UseSMT.java:34)
    at org.kframework.backend.java.symbolic.UseSMTTest.testGetModel(UseSMTTest.java:44)

```

**Comment by cos** [comment]
we could do `mvn clean` instead of `ant clean`

**Comment by cos** [comment]
we could move `lib/maude` to `src/main/maude`

**Comment by grosu** [comment]
Does mvn need ant?  If not, we should then break the dependency on ant entirely, so developers don't even need to install it anymore.

Grigore

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,780 @@
...
+        <groupId>org.codehaus.mojo</groupId>
+        <artifactId>native-maven-plugin</artifactId>
+        <version>1.0-alpha-8</version>
+        <extensions>true</extensions>
+      </plugin>
+      <plugin>
+        <artifactId>maven-antrun-plugin</artifactId>
+        <version>1.7</version>
+        <executions>
+          <execution>
```
Comment (cos) [inline pom.xml:269]: put a comment this is check-style glue

File: pom.xml [hunk 2]
```diff
@@ -0,0 +1,780 @@
...
+              <goal>unpack-dependencies</goal>
+            </goals>
+            <configuration>
+              <includeTypes>zip</includeTypes>
+              <includeScope>runtime</includeScope>
+              <outputDirectory>${project.build.directory}/lib</outputDirectory>
+            </configuration>
+          </execution>
+        </executions>
+      </plugin><!--
```
Comment (cos) [inline pom.xml]: delete comment below

**Comment by cos** [comment]
this is a big step towards cleaning up K. aside for the minor comments above, it looks very good. I agree with the merge.

**Comment by pdaian** [comment]
I skimmed it and it looks good to me, definitely a good start for clean and modular code.  Looking forward to these changes :+1:

**Comment by dwightguth** [comment]
I intend to merge this sometime early next month. It's been pretty well reviewed and I've tested it on several platforms, so I'm pretty sure it works. But not 100%. If, when I merge it, it breaks your code because you didn't review this diff like I asked, I'm not going to take the blame. I'll fix the bug, but I'm not going to take the blame. So please, if you don't want to get blocked, test this pull request on your computer.

**Comment by radumereuta** [comment]
Did you manage to fix the performance issues for `mvn package`?

**Comment by dwightguth** [comment]
`mvn package` on an incremental run on my dev machine now takes about 5 seconds. if you skip unit tests that goes down to about 4 seconds. Running on my laptop, which is a much slower machine, an incremental run now takes 12 seconds if you skip tests. I really can't do much better than this. What remains are bottlenecks in maven plugins that we use which I can't really get around. This is still a huge performance improvement over when you tried it, though. So try it again and let me know what you think.

**Comment by radumereuta** [comment]
Please update src/README
I don't know what are the commands now to build the tool. Everything was explained pretty well in the developers README file.  It would be nice if you could describe the required steps to do the most common tasks:
- download the repository
- build the system (with and w/o tests, clean, partial clean)
- generate eclipse, intellij projects
- run junit tests
- run all tests

You could also describe the overall placement of files, and how they are built, what path to put in your system. It seems that all of these things changed.

**Comment by dwightguth** [comment]
Closing. Will reopen on master
